### PR TITLE
auto-upgrade

### PR DESCRIPTION
**Pull request by clux**
For #262 

Basically exposes a `shipcat self-upgrade` command. This will fetch a specified/latest tag via github releases, unpack the tarball in a temp directory, and swap over the binary in a 3-way swap (potentially screwing up a `brew` installation). It will even swap out the binary at the symlink position if it's linked to somewhere (so people building locally might find themselves with a downloaded release at  `./target/debug/shipcat` - which is probably fine (confused usage)).

When `SHIPCAT_AUTOUPGRADE` evar is set, `shipcat` will invoke `self-upgrade` internally when the version found in shipcat.conf is newer than what we have.

Not tested for Darwin. Need eyes on this please.

### CODE REVIEW

**Comment by clux** [comment]
Because github api rate limits from a single ip, i've added the option to read a github api token optionally under `SHIPCAT_AUTOUPGRADE_TOKEN`. It can have the same permissions as a brew token (repo read).

File: shipcat_cli/src/main.rs [hunk 1]
```diff
@@ -465,8 +472,17 @@ fn resolve_config(args: &ArgMatches, ct: ConfigType) -> Result<(Config, Region)>
         if args.is_present("strict-version-check") {
             return Err(e.into())
         } else {
-            warn!("shipcat version less than pinned minimum - results may vary");
-            warn!("{}", e);
+            if let Ok(_) = std::env::var("SHIPCAT_AUTOUPGRADE") {
```
Comment (janslow) [inline shipcat_cli/src/main.rs]: This can be flattened into an `else if let Ok(_) = std::env::var("SHIPCAT_AUTOUPGRADE")`

Comment (clux) [reply shipcat_cli/src/main.rs]: Ah, thanks. Had forgotten.

**Comment by clux** [comment]
Have tested this with brew. We need to tell people to uninstall it from brew for this to work. Any `brew upgrade shipcat` calls will fail because we overwrite the symlink `brew` manages.

**Comment by clux** [comment]
Merging it for now. It's a feature flagged thing anyway.

**Pull request merged**